python/lib/communication/dmod/communication/client.py

```diff
@@ -504,6 +504,197 @@ def is_new_session(self) -> Optional[bool]:
+        self._active_mux_clients: Dict[str, MuxStreamClient] = dict()
+        """ Clients spawned by ::method:`spawn_mux_client` and not yet retired with ::method:`retire_mux_client`. """
+        self._perform_recv_lock: asyncio.Lock = asyncio.Lock()
+        """ A lock to avoid getting async things out of order due to poorly time queuing in ::method:`mux_recv`. """
+        self._int_mux_id = -1
+        """ Int basis for str mux ids, initially ``-1`` so later logic increments and uses ``0`` as 1st base value. """
+
+        self._len_mux_id = len_mux_ids
+
+    async def mux_recv(self, mux_id: str) -> str:
+        """
+        Receive and return data for a particular mux stream from server over a multiplexed connection.
+
+        Parameters
+        ----------
+        mux_id : str
+            A unique identifier for the specific isolated data stream to receive from.
+
+        Returns
+        -------
+        str
+            The data received from the server, as a string.
+        """
+        while True:
+            # Lock the connection to avoid trouble if two consecutive messages are received AFTER we check the queue,
+            # but a different function call beats "this" one's async await to the first message (and then "this" one
+            # end up with the second message out of order)
+            async with self._perform_recv_lock:
+                # First check if we already have something
+                if self._recv_queues.get(mux_id, None):
+                    return self._recv_queues[mux_id].pop(0)
+
+                # Then try receiving
+                mux_encoded_data = await self._wrapped_client.async_recv()
+                received_mux_id, data = mux_encoded_data[:self._len_mux_id], mux_encoded_data[self._len_mux_id:]
+
+                if mux_id == received_mux_id:
+                    return data
+                elif received_mux_id in self._recv_queues:
+                    self._recv_queues[received_mux_id].append(data)
+                else:
+                    # FIXME: is this the best thing to do?  Should we instead just throw this out?  Or save elsewhere?
+                    msg = f"{self.__class__.__name__} received data on unexpected mux stream {received_mux_id}"
+                    raise RuntimeError(msg)
+
+    async def mux_send(self, data: Union[str, bytearray, bytes], mux_id: str):
+        """
+        Send data to server on a particular mux stream.
+
+        Parameters
+        ----------
+        data: Union[str, bytearray, bytes]
+            The data to send.
+        mux_id : str
+            A unique identifier for an isolated data stream.
+        """
+        if len(mux_id) != self._len_mux_id:
+            msg = f"{self.__class__.__name__} can't send using mux id {mux_id} (expected length {self._len_mux_id!s})"
+            raise ValueError(msg)
+        await self._wrapped_client.async_send(f"{mux_id}{data}")
```

Comment:
It seems that we will need to make changes to the request service's deserialization and serialization behavior to accommodate this change, right? One fear that I have is that this custom serialization format will spread across the entire codebase. For example, if you wanted to use this client and communicate with a particular service directly, that service would have to implement the custom deserialization format and handle stream ordering too.